erasurecoding/src/lib.rs

```diff
@@ -0,0 +1,600 @@
+use libc::{c_int, size_t};
+use std::slice;
+
+const MAX_SHARDS: usize = 65535;
+
+/// Reed-Solomon encode function for FFI
+/// Takes sharded original data as byte array and produces recovery shards
+/// Parameters:
+/// - original_shards: number of original data shards
+/// - original_shards_len: length of original_shards
+/// - recovery_shards: number of recovery shards to generate
+/// - original_shards: input byte array containing flattened shards
+/// - shard_size: size of each shard in bytes
+/// - recovery_shards_out: buffer to store generated and flattened recovery shards
+/// - recovery_shards_out_len: length of recovery_shards_out
+/// Returns 0 on success, -1 on error
+#[no_mangle]
+pub unsafe extern "C" fn reed_solomon_encode(
+    original_shards_count: size_t,
+    recovery_shards_count: size_t,
+    shard_size: size_t,
+    original_shards: *const u8,
+    original_shards_len: size_t,
+    recovery_shards_out: *mut u8,
+    recovery_shards_out_len: size_t,
+) -> c_int {
+    if let Some(sum) = original_shards_count.checked_add(recovery_shards_count) {
+        if sum > MAX_SHARDS {
+            return -1;
+        }
+    } else {
+        // Overflow.
+        return -1;
+    }
```

Comment:
This could be a bit shorter

```suggestion
    match original_shards_count.checked_add(recovery_shards_count) {
        Some(sum) if sum <= MAX_SHARDS => {},
        _ => return -1,
    };
```